okio-assetfilesystem/src/main/kotlin/okio/assetfilesystem/AssetFileSystem.kt

```diff
@@ -0,0 +1,156 @@
+import okio.FileSystem
+import okio.Path
+import okio.Path.Companion.toPath
+import okio.Sink
+import okio.Source
+import okio.source
+
+/**
+ * Expose this [AssetManager] as an Okio [FileSystem].
+ *
+ * Note: Assets are a read-only view on a file system and so any attempt to mutate
+ * will throw an [UnsupportedOperationException].
+ */
+fun AssetManager.asFileSystem(): FileSystem = AssetFileSystem(this)
+
+private class AssetFileSystem(
+  private val assets: AssetManager,
+) : FileSystem() {
+  override fun canonicalize(path: Path): Path {
+    val canonical = canonicalizeInternal(path)
+    if (canonical.existsInternal()) {
+      return canonical
+    }
+    throw FileNotFoundException("$path")
+  }
+
+  private fun canonicalizeInternal(path: Path) = ROOT.resolve(path, normalize = true)
+
+  private fun Path.toAssetRelativePathString(): String {
+    return toString().removePrefix("/")
+  }
+
+  /**
+   * Determine if [this] is a valid path to a file or directory.
+   *
+   * If this function returns true, a call to [AssetManager.open] will either return successfully
+   * or throw [FileNotFoundException] based on whether [this] is a file or directory, respectively.
+   */
+  private fun Path.existsInternal(): Boolean {
+    if (this == ROOT) {
+      return true
+    }
+
+    // Both non-existent paths and paths to existing files return an empty array when listing.
+    // Determine if a path exists by checking if its name is present in the parent's list.
+    val parent = checkNotNull(parent) { "Path has no parent. Did you canonicalize? $this" }
+    val children = assets.list(parent.toAssetRelativePathString()).orEmpty()
+    return name in children
+  }
+
+  override fun metadataOrNull(path: Path): FileMetadata? {
+    val canonical = canonicalizeInternal(path)
+    if (canonical.existsInternal()) {
+      val pathString = canonical.toAssetRelativePathString()
+      return try {
+        assets.open(pathString).close()
+        FileMetadata(
+          isRegularFile = true,
+          isDirectory = false,
+        )
+      } catch (_: FileNotFoundException) {
+        FileMetadata(
+          isRegularFile = false,
+          isDirectory = true,
+        )
+      }
+    }
+    return null
+  }
+
+  override fun list(dir: Path): List<Path> {
+    val canonical = canonicalizeInternal(dir)
+    if (canonical.existsInternal()) {
+      val pathString = canonical.toAssetRelativePathString()
+      try {
+        // This will throw if the path points to a file.
+        assets.open(pathString).close()
+      } catch (_: FileNotFoundException) {
+        return assets.list(pathString)
+          ?.map { it.toPath() }
+          .orEmpty()
+      }
+    }
+    throw FileNotFoundException("$dir")
+  }
+
+  override fun listOrNull(dir: Path): List<Path>? {
+    return try {
+      list(dir)
+    } catch (_: IOException) {
+      null
+    }
+  }
+
+  override fun openReadOnly(file: Path): FileHandle {
+    TODO("not implemented yet!")
```

Comment:
Should do

Reply:
When I wrote this however many years ago I'm pretty sure the resource file system didn't support it so I also decided not to. I see that has since changed, so I'll do an impl as well.

Reply:
I'm going to do this in a follow-up because it's much more complicated than I expected. Assets do not support random access so I've had to fake it.